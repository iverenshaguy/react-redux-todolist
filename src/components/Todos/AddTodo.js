import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddTodo extends Component {
  constructor() {
    super();
    this.state = {
      todo: {},
      value: ''
    };
  }

  handleChange = event => {
    this.setState({
      value: event.target.value
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    if (this.state.value === '') {
      alert('Title is required');
      return;
    }

    if (this.props.addTodo) {
      this.setState(
        {
          todo: {
            userId: 1,
            id:
              this.props.todos.length >= 1
                ? +this.props.todos[this.props.todos.length - 1].id + 1
                : 1,
            title: this.state.value,
            completed: false
          }
        },
        () => {
          this.props.addTodo(this.state.todo);
        }
      );
    } else if (this.props.editTodo) {
      const editedTodo = {
        userId: 1,
        id: this.props.todo.id,
        title: this.state.value,
        completed: false
      };
      this.setState(
        prevState => ({
          todo: Object.assign({}, this.state.todo, editedTodo)
        }),
        () => {
          this.props.editTodo(this.state.todo);
        }
      );
    }
  };

  render() {
    return (
      <div className="AddTodo">
        <form className="form-inline my-2 my-lg-0" onSubmit={this.handleSubmit}>
          <input
            className="form-control mr-sm-2 new-todo"
            type="text"
            name="todoinput"
            value={this.state.value}
            onChange={this.handleChange}
            placeholder="Enter Todo"
            autoFocus={true}
          />
          <button type="submit" className="btn btn-primary my-2 my-sm-0">
            {this.props.buttonText}
          </button>
        </form>
      </div>
    );
  }
}

AddTodo.proptypes = {
  todo: PropTypes.object,
  addTodo: PropTypes.func,
  editTodo: PropTypes.func,
  buttonText: PropTypes.string.isRequired
};

export default AddTodo;
