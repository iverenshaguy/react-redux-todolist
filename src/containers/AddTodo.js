import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import uuidv4 from 'uuid/v4';
import { addTodo, editTodo, toggleEditTodo } from '../store/Todos/actions';

let AddTodo = props => {
  let input;

  const dispatch = props.dispatch;

  return (
    <div>
      <form
        className="form-inline my-2 my-lg-0"
        onSubmit={e => {
          e.preventDefault();
          if (!input.value.trim()) {
            alert('Title is required');
            return;
          }

          if (props.addTodo) {
            const todo = {
              id: uuidv4(),
              title: input.value,
              completed: false
            };

            dispatch(addTodo(todo));
          } else if (props.editTodo) {
            const editedTodo = {
              id: props.todo.id,
              title: input.value,
              completed: false
            };

            dispatch(editTodo(editedTodo));
            dispatch(toggleEditTodo(props.todo.id));
          }

          input.value = '';
        }}
      >
        <input
          className="form-control mr-sm-2 new-todo"
          type="text"
          name="todoinput"
          ref={node => {
            input = node;
          }}
          placeholder="Enter Todo"
          autoFocus={true}
        />
        <button type="submit" className="btn btn-primary my-2 my-sm-0">
          {props.buttonText}
        </button>
      </form>
    </div>
  );
};

AddTodo.proptypes = {
  todo: PropTypes.object,
  addTodo: PropTypes.bool,
  editTodo: PropTypes.bool,
  buttonText: PropTypes.string.isRequired
};

AddTodo = connect()(AddTodo);

export default AddTodo;
