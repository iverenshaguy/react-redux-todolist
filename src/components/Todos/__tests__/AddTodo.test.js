import AddTodo from '../AddTodo';
import { todos as mockTodos } from '../../../__mocks__/todos.mock';

const setup = () => {
  const props = {
    buttonText: 'Add Todo'
  };
  const shallowWrapper = shallow(<AddTodo {...props} />);
  const mountedWrapper = mount(<AddTodo {...props} />);

  return {
    shallowWrapper,
    mountedWrapper,
    props
  };
};

describe('AddTodo', () => {
  it('renders correctly', () => {
    const { shallowWrapper } = setup();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('mounts correctly', () => {
    const { mountedWrapper } = setup();
    expect(toJson(mountedWrapper)).toMatchSnapshot();
  });

  it('responds to todo change', () => {
    const event = { target: { value: 'Go to Church' } };
    const { mountedWrapper } = setup();
    // const handleChangeSpy = jest.spyOn(
    //   mountedWrapper.instance(),
    //   'handleChange'
    // );
    expect(toJson(mountedWrapper)).toMatchSnapshot();
    mountedWrapper.find('input').simulate('change', event);
    mountedWrapper
      .find('input')
      .simulate('keyDown', { keyCode: 9, key: 'Tab' });
    // expect(handleChangeSpy).toHaveBeenCalledTimes(1);
    expect(mountedWrapper.instance().state.value).toBe('Go to Church');
  });

  it('alerts warning for empty todo input', () => {
    const event = { target: { value: '' } };
    const { mountedWrapper } = setup();
    window.alert = jest.fn();
    mountedWrapper.find('input').simulate('change', event);
    mountedWrapper.find('form').simulate('submit');
    expect(window.alert).toHaveBeenCalledWith('Title is required');
  });

  it('adds new todo to state', () => {
    const event = { target: { value: 'Read My Bible' } };
    const { mountedWrapper } = setup();
    mountedWrapper.find('input').simulate('change', event);
    mountedWrapper.find('form').simulate('submit');
    expect(mountedWrapper.instance().state.value).toBe('Read My Bible');
  });

  it('creates a new todo object and sends it Todos with empty todos array', () => {
    const event = { target: { value: 'Wash my Clothes' } };
    const response = {
      userId: 1,
      id: 1,
      title: 'Wash my Clothes',
      completed: false
    };

    const addTodo = jest.fn();
    const wrapper = mount(
      <AddTodo {...setup().props} addTodo={addTodo} todos={[]} />
    );

    wrapper.find('input').simulate('change', event);
    wrapper.find('form').simulate('submit');
    expect(wrapper.instance().state.todo).toEqual(response);
  });

  it('creates a new todo object and sends it Todos', () => {
    const event = { target: { value: 'Learn React TDD' } };
    const response = {
      userId: 1,
      id: 5,
      title: 'Learn React TDD',
      completed: false
    };

    const addTodo = jest.fn();
    const wrapper = mount(
      <AddTodo {...setup().props} addTodo={addTodo} todos={mockTodos} />
    );

    wrapper.find('input').simulate('change', event);
    wrapper.find('form').simulate('submit');
    expect(wrapper.instance().state.todo).toEqual(response);
  });

  it('edits a todo object', () => {
    const event = { target: { value: 'Learn React Test Driven Development' } };
    const response = {
      userId: 1,
      id: 3,
      title: 'Learn React Test Driven Development',
      completed: false
    };

    const addTodo = jest.fn();
    const wrapper = mount(
      <AddTodo {...setup().props} editTodo={addTodo} todo={mockTodos[2]} />
    );

    wrapper.find('input').simulate('change', event);
    wrapper.find('form').simulate('submit');
    expect(wrapper.instance().state.todo).toEqual(response);
  });
});
