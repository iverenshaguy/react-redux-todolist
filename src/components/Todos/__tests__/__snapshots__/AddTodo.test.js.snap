// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo mounts correctly 1`] = `
<AddTodo
  buttonText="Add Todo"
>
  <div
    className="AddTodo"
  >
    <form
      className="form-inline my-2 my-lg-0"
      onSubmit={[Function]}
    >
      <input
        autoFocus={true}
        className="form-control mr-sm-2 new-todo"
        name="todoinput"
        onChange={[Function]}
        placeholder="Enter Todo"
        type="text"
        value=""
      />
      <button
        className="btn btn-primary my-2 my-sm-0"
        type="submit"
      >
        Add Todo
      </button>
    </form>
  </div>
</AddTodo>
`;

exports[`AddTodo renders correctly 1`] = `
<div
  className="AddTodo"
>
  <form
    className="form-inline my-2 my-lg-0"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="form-control mr-sm-2 new-todo"
      name="todoinput"
      onChange={[Function]}
      placeholder="Enter Todo"
      type="text"
      value=""
    />
    <button
      className="btn btn-primary my-2 my-sm-0"
      type="submit"
    >
      Add Todo
    </button>
  </form>
</div>
`;

exports[`AddTodo responds to todo change 1`] = `
<AddTodo
  buttonText="Add Todo"
>
  <div
    className="AddTodo"
  >
    <form
      className="form-inline my-2 my-lg-0"
      onSubmit={[Function]}
    >
      <input
        autoFocus={true}
        className="form-control mr-sm-2 new-todo"
        name="todoinput"
        onChange={[Function]}
        placeholder="Enter Todo"
        type="text"
        value=""
      />
      <button
        className="btn btn-primary my-2 my-sm-0"
        type="submit"
      >
        Add Todo
      </button>
    </form>
  </div>
</AddTodo>
`;
